src/profilers.cpp

```diff
@@ -2,19 +2,369 @@
+            }
+          }else{
+            parsed_data = data;
+          }
+
+          std::regex number_regex(R"(-?\d+(\.\d+)?)");;
+
+          std::smatch match;
+          std::string temp_data = parsed_data;
+          int count = 0;
+          float metric_val;
+                    
+          while (std::regex_search(temp_data, match, number_regex)) {
+            count++;
+            if(count > 1){break;}
+  
+            std::string number_str = match.str(0);
+              try {
+                  metric_val = std::stof(number_str);  
+              } catch (const std::exception&) {
+                    // failed conversion need to handle later
+                    std::exit(ERROR_CONVERSION_TO_FLOAT);
+              }
+            temp_data = match.suffix().str();
+          
+          }
+          if(count == 0){
+            //no numbers found 
+            std::exit(ERROR_NO_NUMBER_REGEX);
+          }
+          if(count > 1){
+            //error too many numbers
+            std::exit(ERROR_TOO_MANY_NUMBERS_REGEX);
+          }
+
+          close(pipe_fd[0]);
+
+
+          // ["<CUSTOM_METRIC>", <metric-reading command string>, <user-provided metric name string>, <timestamp in nanoseconds>, <value of the metric>]
+          clock_gettime(CLOCK_MONOTONIC, &end);
+          long timestamp = (end.tv_sec - start.tv_sec) * 1000 * 1000000 + (end.tv_nsec - start.tv_nsec);
+          nlohmann::json metric = nlohmann::json::array({"<CUSTOM_METRIC>", this->metric_command , this->metric_name, timestamp, metric_val });
+
+          std::string s = metric.dump();
+
+
+          //connection write s
+          std::unique_ptr<Connection> connection;
+
+          std::regex pipe_regex(R"(pipe\s+(\d+)_(\d+))");
+          std::regex tcp_regex(R"(TCP\s+([\d\.]+)_(\d+))");
+          std::smatch instruction_match;
+
+          if (std::regex_match(instrs, instruction_match, pipe_regex)) {
+              
+              int read_fd[2];
+              int write_fd[2];
+
+              write_fd[1] = std::stoi(match[1].str());
+              read_fd[0] = std::stoi(match[2].str());
+              connection = std::make_unique<FileDescriptor>(write_fd, read_fd, this->server_buffer);
+          } else if (std::regex_match(instrs, instruction_match, tcp_regex)) {
+              
+              std::string server_address = match[1].str() + ":" + match[2].str();
+              
+
+              Poco::Net::SocketAddress address(server_address);
+              Poco::Net::StreamSocket socket(address);
+
+              connection = std::make_unique<TCPSocket>(socket, this->server_buffer);
+              
+          }else{
+              // instruction is not correct
+              std::exit(ERROR_PARSING_CONNECTION_INSTRS);
+          }
+
+          connection->write(s);
+        }
+
+        close(pipe_fd[1]);
+        close(pipe_fd[0]);
+
+        command_status = waitpid(pid, &status, WNOHANG); //check if command to be profiled is still executing
```

Comment:
I'd move it to the very end of the loop (after ```nanosleep```).

Reply:
Done :+1: